src/draive/splitters/basic.py

```diff
@@ -33,74 +45,71 @@ def split_text(
 
 def _split(
     text: str,
-    separators: tuple[str, str] | str | None = None,
-) -> tuple[str, list[str]]:
-    # prepare initial splitter - default is new paragraph
-    splitter: str
-    alt_splitter: str
-    match separators:
-        case (str(primary), str(secondary)):
-            splitter = primary
-            alt_splitter = secondary
-        case str(primary):
-            splitter = primary
-            alt_splitter = "\n"
-        case None:
-            splitter = "\n\n"
-            alt_splitter = "\n"
+    splitters: Sequence[str],
+) -> tuple[str, list[str] | None, list[str]]:
+    # try splitting using provided splitters
+    for index, splitter in enumerate(splitters):
+        parts: list[str] = text.split(splitter)
+        # if splitting has done nothing retry using secondary splitter
+        if len(parts) == 1:
+            continue
+        else:
+            remained_splitters: list[str] | None
+            next_index = index + 1
+            if next_index < len(splitters):
+                remained_splitters = list(splitters[next_index:])
+            else:
+                remained_splitters = None
 
-    # try splitting using provided splitter
-    parts: list[str] = text.split(splitter)
-    # if splitting has done nothing retry using secondary splitter
-    if len(parts) == 1:
-        splitter = alt_splitter
-        parts = text.split(splitter)
+            return (splitter, remained_splitters, parts)
     # if splitting has still done nothing then fail
-    if len(parts) == 1:
-        raise ValueError("Failed to properly split text with provided separators")
+    raise ValueError("Failed to properly split text with provided separators")
 
-    return (splitter, parts)
 
-
-def _merge(  # noqa: C901
+def _merge(  # noqa: C901, PLR0912, PLR0913
     parts: list[str],
     part_size: int,
     count_size: Callable[[str], int],
     splitter: str,
+    remained_splitters: list[str] | None,
     part_overlap_size: int | None,
 ) -> list[str]:
     result: list[str] = []
     accumulator: list[str] = []
     accumulator_size: int = 0
     # iterate over splitted pats
     for part in parts:
-        temp_size: int = count_size(part)
+        current_part: str = part
+        # Add the separator back if it's not the first part
+        if accumulator:
+            current_part = splitter + part
+
+        temp_size: int = count_size(current_part)
```

Comment:
I am starting to think that we should count the size of joined thing instead of adding numbers to be more accurate 🤔